File: crabzilla-core/src/main/java/io/github/crabzilla/core/entity/EntityUnitOfWork.java
package io.github.crabzilla.core.entity;

import io.github.crabzilla.core.DomainEvent;
import lombok.NonNull;
import lombok.Value;

import java.io.Serializable;
import java.util.List;
import java.util.UUID;

@Value
public class EntityUnitOfWork implements Serializable {

  private UUID unitOfWorkId;
  private EntityCommand command;
  private Version version;
  private List<DomainEvent> events;

  public static EntityUnitOfWork unitOfWork(@NonNull EntityCommand command,
                                            @NonNull Version version,
                                            @NonNull List<DomainEvent> events) {
    return new EntityUnitOfWork(UUID.randomUUID(), command, version, events);
  }

  public EntityId targetId() {
    return command.getTargetId();
  }

}

File: crabzilla-core/src/main/java/io/github/crabzilla/core/entity/EntityUnitOfWork.kt
package io.github.crabzilla.core.entity

import com.fasterxml.jackson.annotation.JsonTypeInfo
import io.github.crabzilla.core.Command
import io.github.crabzilla.core.DomainEvent
import java.io.Serializable
import java.util.*

@JsonTypeInfo(use = JsonTypeInfo.Id.CLASS, include = JsonTypeInfo.As.PROPERTY, property = "@class")
interface UnitOfWork {
  val unitOfWorkId: UUID
  val command: Command
  val version: Version
  val events: List<DomainEvent>
  fun targetId(): EntityId
}

data class EntityUnitOfWork(override val unitOfWorkId: UUID, override val command: EntityCommand,
                            override val version: Version, override val events: List<DomainEvent>) : UnitOfWork {

  override fun targetId(): EntityId {
    return command.targetId
  }
}

data class ExternalServiceUnitOfWork(val id: EntityId, override val unitOfWorkId: UUID, override val command: Command,
                                     override val version: Version, override val events: List<DomainEvent>) : UnitOfWork {
  override fun targetId(): EntityId {
    return id
  }

}
